### PR TITLE
[Trainer] Make sure shown loss in distributed training is correctly averaged over all workers

### PR DESCRIPTION
**Pull request by patrickvonplaten**
# What does this PR do?

Currently, the only the loss of the first workes is shown in distributed training. However I think it would be cleaner to show the averaged loss over all workers. In some setups (CTC speech recognition e.g.) the loss can vary quite a bit from worker to worker.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/master/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/master/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/master/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

**Comment by patrickvonplaten** [comment]
What do you think @sgugger , @stas00 ? Any good ideas of how to write a test for it?

File: src/transformers/trainer.py [hunk 1]
```diff
@@ -1834,7 +1834,15 @@ def training_step(self, model: nn.Module, inputs: Dict[str, Union[torch.Tensor,
...
             loss.backward()
 
-        return loss.detach()
+        loss = loss.detach()
+
+        # make sure that we display the averaged loss over all workers
+        # in distributed mode
+        if self.args.world_size > 1:
+            dist.reduce(loss)
+            loss /= self.args.world_size
```
[the remark this answers is not in the document]
Comment (patrickvonplaten) [reply src/transformers/trainer.py:1843]: Good point!

**Review by stas00 — COMMENTED** [review]
That's a good point, @patrickvonplaten 

This would made things inconsistent, since all other metrics are for rank 0 and aren't averaged or even measured.

Would it be too noisy if we added both rank0 and average loss? 

for visual reporting it can even be a single line:
```
loss    = rank0: 2.33 average: 2.30
```
for normal data log these should be 2 different points. Also concerned with BC, since it may now report different results potentially and break user's workflow. So possibly leaving `loss` as rank0 and adding `loss_average` for average? not attached to the naming.

In the future perhaps we should review and see if there are other metrics that we too should average. In which case we could add a toggle flag which would report rank0 or average? And then a user activating this toggle will not have a BC issue if the default remains `rank0`

**Comment by sgugger** [comment]
> This would made things inconsistent, since all other metrics are for rank 0 and aren't averaged or even measured.

All metrics (apart from the speed/RAM metrics) are computed on gathered predictions and labels, so this would actually make things more consistent. The validation loss for instance is the validation loss across all processes.

**Comment by stas00** [comment]
Oh, if it's just speed/mem metrics, then of course go for it. I wonder if we should then flag in the reports that speed/mem metrics are for gpu0. But we can tackle it in another PR.

**Comment by patrickvonplaten** [comment]
Verified that the new line works for distributed training. Should I add a test here @sgugger ?

File: src/transformers/trainer.py [hunk 2]
```diff
@@ -1462,7 +1462,10 @@ def _load_state_dict_in_model(self, state_dict):
     def _maybe_log_save_evaluate(self, tr_loss, model, trial, epoch, ignore_keys_for_eval):
         if self.control.should_log:
             logs: Dict[str, float] = {}
-            tr_loss_scalar = tr_loss.item()
+
+            # all_gather + mean() to get average loss over all processes
+            tr_loss_scalar = self._nested_gather(tr_loss).mean().item()
```
Comment (sgugger) [inline src/transformers/trainer.py:1467]: Very clean!
Is it properly working without putting it in a 1d tensor? It's always tricky with gather.

**Review by sgugger — APPROVED** [review]

File: src/transformers/trainer.py (hunk 2 above)
Comment (patrickvonplaten) [reply src/transformers/trainer.py:1467]: I ran a dummy training script for distributed training and the training logs were displayed, so I think it works correctly for a scalar tensor. 

Maybe you can run this script in a distributed setup - e.g. Vorace wiht 4 GPUs - to make sure everything works as expected:

```bash
python -m torch.distributed.launch \
        --nproc_per_node 4 run_speech_recognition_ctc.py \
        --dataset_name="common_voice" \
        --model_name_or_path="facebook/wav2vec2-base" \
        --dataset_config_name="ab" \
        --output_dir="./dummy" \
        --preprocessing_num_workers="1" \
        --overwrite_output_dir \
        --num_train_epochs="2" \
        --per_device_train_batch_size="1" \
        --gradient_accumulation_steps="1" \
        --learning_rate="3e-4" \
        --warmup_steps="500" \
        --evaluation_strategy="steps" \
        --audio_column_name="path" \
        --text_column_name="sentence" \
        --save_steps="400" \
        --eval_steps="100" \
        --logging_steps="1" \
        --layerdrop="0.0" \
        --save_total_limit="3" \
        --freeze_feature_extractor \
        --chars_to_ignore , ? . ! - \; \: \" “ % ‘ ” � \
        --fp16 \
        --group_by_length \
        --do_train --do_eval
```

Comment (patrickvonplaten) [reply src/transformers/trainer.py:1467]: with `run_speech_recognition_ctc.py` being here: https://github.com/huggingface/transformers/blob/master/examples/pytorch/speech-recognition/run_speech_recognition_ctc.py

Comment (sgugger) [reply src/transformers/trainer.py:1467]: Tested locally and it works perfectly indeed. Thanks!
